### PR TITLE
fix(pbts): legacy `ABCIParams` values are properly translated to `FeatureParams` values

### LINKED ISSUE #2438: State store should be able to unmarshal the old version of ConsensusParams protobuf from the DB
V1 introduces FeatureParams and deprecates ABCIParams, as fields of ConsensusParams.
As part of the change, `ABCIParams.vote_extensions_enabled_height` was replaced by `FeatureParams.vote_extensions_enabled_height`

This issue is open to ensure that networks upgrading from v0.38 to v1 and which have set an enabling height before the migration, do the right thing and move the value from one field to the other.

### PR DESCRIPTION
**Pull request by lasarojc**
Closes #2438

---

#### PR checklist

- [x] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

**Comment by lasarojc** [comment]
Have confirmed the correct behavior of this fix manually, with the following runs/results

1. Start the chain with v0.38.x, with enable height set to X
   - Upgrade to v1 (with fix) before X
      - legacy value X is detected
      - Consensus parameters properly set (FeatureParams.vote_extensions_enabled_height = 100)
   - Restart the node before X
      - legacy value is not detected
      - Consensus parameters properly set (FeatureParams.vote_extensions_enabled_height = 100)
   - VE are enabled at X
   - Restart the node after X
      - legacy value is not detected
      - Consensus parameters properly set (FeatureParams.vote_extensions_enabled_height = 100)
2. Start the chain with v0.38.x, with enable height set to X
   - Upgrade to v1 (with fix) before X
      - legacy value X is detected
      - Consensus parameters properly set (FeatureParams.vote_extensions_enabled_height = 100)
   - VE are enabled at X
   - Restart the node after X
      - legacy value is not detected
      - Consensus parameters properly set (FeatureParams.vote_extensions_enabled_height = 100)
3. Start the chain with v0.38.x, with enable height set to X
   - VE are enabled at X
   - Restart the node after X
      - legacy value is not detected
      - Consensus parameters properly set (FeatureParams.vote_extensions_enabled_height = 100)

File: types/params.go [hunk 1]
```diff
@@ -463,22 +463,24 @@ func ConsensusParamsFromProto(pbParams cmtproto.ConsensusParams) ConsensusParams
 		Version: VersionParams{
 			App: pbParams.Version.App,
 		},
+		Feature: FeatureParams{
+			VoteExtensionsEnableHeight: pbParams.GetFeature().GetVoteExtensionsEnableHeight().GetValue(),
+			PbtsEnableHeight:           pbParams.GetFeature().GetPbtsEnableHeight().GetValue(),
+		},
```
Comment (lasarojc) [inline types/params.go:469]: This block replaces lines 467-473 on the left

File: types/params.go [hunk 2]
```diff
@@ -463,22 +463,24 @@ func ConsensusParamsFromProto(pbParams cmtproto.ConsensusParams) ConsensusParams
...
 	}
-	if pbParams.Synchrony != nil {
-		if pbParams.Synchrony.MessageDelay != nil {
-			c.Synchrony.MessageDelay = *pbParams.Synchrony.GetMessageDelay()
-		}
-		if pbParams.Synchrony.Precision != nil {
-			c.Synchrony.Precision = *pbParams.Synchrony.GetPrecision()
+	if pbParams.GetSynchrony().GetPrecision() != nil {
+		c.Synchrony.Precision = *pbParams.GetSynchrony().GetPrecision()
+	}
```
Comment (lasarojc) [inline types/params.go:476]: This block replaces lines 474-481 on the left.

File: types/params.go [hunk 3]
```diff
@@ -463,22 +463,24 @@ func ConsensusParamsFromProto(pbParams cmtproto.ConsensusParams) ConsensusParams
...
+	if pbParams.GetAbci().GetVoteExtensionsEnableHeight() > 0 {
+		// Value set before the upgrade to V1. We can safely overwrite here because
+		// ABCIParams and FeatureParams being set is mutually exclusive (<V1 and >=V1).
+		if pbParams.GetFeature().GetVoteExtensionsEnableHeight().GetValue() > 0 {
+			panic("vote_extension_enable_height is set in two different places")
 		}
+		c.Feature.VoteExtensionsEnableHeight = pbParams.Abci.VoteExtensionsEnableHeight
```
Comment (lasarojc) [inline types/params.go:483]: This block actually converts the legacy value into the new value.

File: test/e2e/networks/upgrade.toml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+# Image cometbft/v0.38.x built with v0.38.x as of feb 27/2024
+
+[initial_state]
+
+[validators]
+  validator01 = 50
+  validator02 = 50
+  validator03 = 50
+  validator04 = 50
```
Comment (sergio-mena) [inline test/e2e/networks/upgrade.toml:17]: Do you really need this? ... anyway up to you if you wanna keep it

File: test/e2e/networks/upgrade.toml [hunk 5]
```diff
@@ -0,0 +1,44 @@
+ipv6 = false
+initial_height = 0
```
Comment (sergio-mena) [inline test/e2e/networks/upgrade.toml:2]:
```suggestion
```

File: test/e2e/networks/upgrade.toml [hunk 6]
```diff
@@ -0,0 +1,44 @@
+ipv6 = false
```
Comment (sergio-mena) [inline test/e2e/networks/upgrade.toml:1]:
```suggestion
ipv6 = false
```
default value, no need to appear

File: test/e2e/networks/upgrade.toml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+initial_height = 0
+key_type = ""
```
Comment (sergio-mena) [inline test/e2e/networks/upgrade.toml:3]:
```suggestion
```

File: test/e2e/networks/upgrade.toml [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+evidence = 10
+vote_extensions_enable_height = 100
+abci_protocol = "tcp"
+upgrade_version = "cometbft/e2e-node:local-version"
+
+# Image cometbft/v0.38.x built with v0.38.x as of feb 27/2024
+
+[initial_state]
+
```
Comment (sergio-mena) [inline test/e2e/networks/upgrade.toml:12]:
```suggestion
```

**Review by sergio-mena — APPROVED** [review]
Excellent!

**Pull request merged**